api/client/cp.go

```diff
@@ -10,48 +15,242 @@ import (
+}
+
+func resolveLocalPath(localPath string) (absPath string, err error) {
+	if absPath, err = filepath.Abs(localPath); err != nil {
+		return
+	}
+
+	return archive.PreserveTrailingDotOrSeparator(absPath, localPath), nil
+}
+
+func (cli *DockerCli) copyFromContainer(srcContainer, srcPath, dstPath string) (err error) {
+	if dstPath != "-" {
+		// Get an absolute destination path.
+		dstPath, err = resolveLocalPath(dstPath)
+		if err != nil {
+			return err
+		}
+	}
+
+	query := make(url.Values, 1)
+	query.Set("path", filepath.ToSlash(srcPath)) // Normalize the paths used in the API.
+
+	urlStr := fmt.Sprintf("/containers/%s/archive-path?%s", srcContainer, query.Encode())
+
+	response, hdr, statusCode, err := cli.call("GET", urlStr, nil, nil)
 	if err != nil {
 		return err
 	}
+	defer response.Close()
 
-	hostPath := cmd.Arg(1)
-	if statusCode == 200 {
-		if hostPath == "-" {
-			_, err = io.Copy(cli.out, stream)
-		} else {
-			err = archive.Untar(stream, hostPath, &archive.TarOptions{NoLchown: true})
+	if statusCode != 200 {
+		return fmt.Errorf("unexpected status code from daemon: %d", statusCode)
+	}
+
+	if dstPath == "-" {
+		// Send the response to STDOUT.
+		_, err = io.Copy(os.Stdout, response)
+
+		return err
+	}
+
+	var stat types.ContainerPathStat
+	encodedStat := hdr.Get("X-Docker-Container-Path-Stat")
+	statDecoder := base64.NewDecoder(base64.StdEncoding, strings.NewReader(encodedStat))
+	err = json.NewDecoder(statDecoder).Decode(&stat)
+	if err != nil {
+		return fmt.Errorf("unable to decode container path stat header: %s", err)
+	}
+
+	// Prepare source copy info.
+	srcInfo := archive.CopyInfo{
+		Path:   srcPath,
+		Exists: true,
+		IsDir:  stat.Mode.IsDir(),
+	}
+
+	return archive.CopyTo(response, srcInfo, dstPath)
+}
+
+func (cli *DockerCli) copyToContainer(srcPath, dstContainer, dstPath string) (err error) {
+	if srcPath != "-" {
+		// Get an absolute source path.
+		srcPath, err = resolveLocalPath(srcPath)
+		if err != nil {
+			return err
+		}
+	}
+
+	// Prepare destination copy info by stat-ing the container path.
+	dstInfo := archive.CopyInfo{Path: dstPath}
+	dstStat, err := cli.statContainerPath(dstContainer, dstPath)
+	if err == nil {
+		dstInfo.Exists, dstInfo.IsDir = true, dstStat.Mode.IsDir()
+	}
+	// Ignore any other error and assume that the parent directory of the
+	// destination path exists, in which case the copy may still succeed. If
+	// there is any type of conflict (e.g., non-directory overwriting an
+	// existing directory or vice versia) the extraction will fail. If the
+	// destination simply did not exist, but the parent directory does, the
+	// extraction will still succeed.
+
+	var content io.Reader
+	if srcPath == "-" {
+		// Use STDIN.
+		content = os.Stdin
+		if !dstInfo.IsDir {
+			return fmt.Errorf("destination is not a directory")
+		}
+	} else {
+		srcArchive, err := archive.TarResource(srcPath)
+		if err != nil {
+			return err
 		}
+		defer srcArchive.Close()
+
+		// Prepare source copy info.
```

Comment:
@jlhawn can you elaborate on what's going on in this bit of code and why the tar of the src above isn't enough?

If this is twiddling the src archive based on the target, I wonder if it wouldn't be better to do it on the daemon side. In cases where a 3rd party (or user) is trying to call the API directly I think we should make it as easy as possible on them, which probably means just a simple tar of the src files. If we expect them to modify the tar based on the dest then there's a good chance they may get it wrong. How many will have some kind of PrepareArchiveCopy type of func handy?

Also, it seems like what the API expects shouldn't change based on the dest. Meaning, I would expect the incoming tar to be the same irrespective of where its going and leave it up to the daemon (un-tarring logic) to figure out if needs to be tweaked.